### PR TITLE
Editable text component

### PR DESCRIPTION
**Pull request by devrk96**

### CODE REVIEW

File: app/client/src/components/ads/Icon.tsx [hunk 1]
```diff
@@ -81,6 +89,27 @@ export const Icon = (props: IconProps) => {
         </IconWrapper>
       );
       break;
+    case "edit":
```
Comment (riodeuno) [inline app/client/src/components/ads/Icon.tsx:92]: I was wondering if we can get these values from an enum instead?

File: app/client/src/components/ads/EditableText.tsx [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+  };
+
+  return (
+    <EditableTextWrapper
+      fill={props.fill}
+      onMouseEnter={nonEditMode}
+      onDoubleClick={
+        props.editInteractionKind === EditInteractionKind.DOUBLE
+          ? editMode
+          : _.noop
```
Comment (riodeuno) [inline app/client/src/components/ads/EditableText.tsx:252]: Please tree-shake the lodash module wherever possible. `import {noop} from "lodash"`

File: app/client/src/components/ads/EditableText.tsx [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+    }
+  };
+
+  const iconName =
+    !isEditing &&
+    !savingState.isSaving &&
+    savingState.name === SavingState.NOT_STARTED
+      ? "edit"
+      : !isEditing &&
+        savingState.isSaving &&
```
Comment (riodeuno) [inline app/client/src/components/ads/EditableText.tsx:225]: `savingState.isSaving` should be `false` here or `true`? Please revisit this code.
`savingState.isSaving` check may not be required as we're doing the `savingState.name === SavingState.NOT_STARTED` above. Please check this part of the code.

File: app/client/src/components/ads/EditableText.tsx [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+`;
+
+export const AdsEditableText = (props: EditableTextProps) => {
+  const [isEditing, setIsEditing] = useState(!!props.isEditingDefault);
+  const [value, setValue] = useState(props.defaultValue);
+  const [lastValidValue, setLastValidValue] = useState(props.defaultValue);
+  const [isInvalid, setIsInvalid] = useState<string | boolean>(false);
+  const [changeStarted, setChangeStarted] = useState<boolean>(false);
+  const [savingState, setSavingState] = useState<{
+    isSaving: boolean;
```
Comment (riodeuno) [inline app/client/src/components/ads/EditableText.tsx:144]: `isSaving` seems to be a duplicate of `SavingState.STARTED`. Please confirm if we need both these properties in the state.

File: app/client/src/components/ads/EditableText.tsx [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+  useEffect(() => {
+    if (props.forceDefault === true) setValue(props.defaultValue);
+  }, [props.forceDefault, props.defaultValue]);
+
+  const bgColor = useMemo(
+    () => editModeBgcolor(!!isInvalid, isEditing, savingState),
+    [isInvalid, isEditing, savingState],
+  );
+
+  const editMode = (e: React.MouseEvent) => {
```
Comment (riodeuno) [inline app/client/src/components/ads/EditableText.tsx:162]: `editMode`, `onInputChange` may need to be wrapped in `useCallback` if the dependencies change frequently.

**Review by riodeuno — CHANGES_REQUESTED** [review]

**Review by satbir121 — APPROVED** [review]

**Review by satbir121 — APPROVED** [review]

**Review by satbir121 — APPROVED** [review]

**Pull request merged**